mistralrs-core/src/sequence.rs

```diff
@@ -231,20 +235,47 @@ impl Sequence {
         self.state.set(state);
     }
 
-    pub fn is_done(&self, tok: u32, eos_tok: u32, max_model_len: usize) -> Option<StopReason> {
+    pub fn generated_text(&self, tokenizer: &Tokenizer) -> anyhow::Result<String> {
+        let token_ids = self
+            .logprobs()
+            .iter()
+            .map(|lp| lp.token)
+            .collect::<Vec<_>>();
+        let text = tokenizer
+            .decode(&token_ids, false)
+            .map_err(|e| anyhow::Error::msg(format!("Tokenization failure - {}", e)))?;
+
+        Ok(text)
+    }
+
+    pub fn is_done(
+        &self,
+        tok: u32,
+        eos_tok: u32,
+        max_model_len: usize,
+        tokenizer: &Tokenizer,
+    ) -> anyhow::Result<Option<StopReason>> {
         if tok == eos_tok {
-            Some(StopReason::Eos)
+            Ok(Some(StopReason::Eos))
         } else if self.stop_tokens.contains(&tok) {
-            Some(StopReason::StopTok(tok))
+            Ok(Some(StopReason::StopTok(tok)))
         } else if self.max_len.is_some()
             && self.tokens.len().saturating_sub(self.prompt_len) == self.max_len.unwrap()
         {
             // add_token was already called
-            Some(StopReason::Length(self.max_len.unwrap()))
+            Ok(Some(StopReason::Length(self.max_len.unwrap())))
         } else if self.tokens.len().saturating_sub(self.prompt_len) == max_model_len {
-            Some(StopReason::ModelLength(max_model_len))
+            Ok(Some(StopReason::ModelLength(max_model_len)))
         } else {
-            None
+            if !self.stop_strings.is_empty() {
+                let text_acc = self.generated_text(tokenizer)?;
+                for (idx, s) in self.stop_strings.iter().enumerate() {
+                    if text_acc.contains(s) {
+                        return Ok(Some(StopReason::StopString(idx)));
+                    }
+                }
+            }
```

Comment:
I could not think of how to optimize this, besides not running if not requested

Reply:
Yes, I think this is a good solution.